File: 17-Binary-Tree/java/07-BottomViewOfBinaryTree.java
class Solution {
    public List<Integer> bottomView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;

        Map<Integer, Integer> map = new TreeMap<>();
        Queue<Pair> queue = new LinkedList<>();
        queue.offer(new Pair(root, 0));

        while (!queue.isEmpty()) {
            Pair current = queue.poll();
            TreeNode node = current.node;
            int hd = current.hd;

            map.put(hd, node.data);

            if (node.left != null) {
                queue.offer(new Pair(node.left, hd - 1));
            }
            if (node.right != null) {
                queue.offer(new Pair(node.right, hd + 1));
            }
        }

        for (int value : map.values()) {
            result.add(value);
        }

        return result;
    }

    class Pair {
        TreeNode node;
        int hd;

        Pair(TreeNode node, int hd) {
            this.node = node;
            this.hd = hd;
        }
    }
}

File: 17-Binary-Tree/cpp/07-bottom-view-of-binary-tree.cpp
class Solution {
public:
    vector<int> bottomView(TreeNode* root) {
        vector<int> result;
        if (root == nullptr) return result;

        map<int, int> mp;
        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});

        while (!q.empty()) {
            auto current = q.front();
            q.pop();

            TreeNode* node = current.first;
            int hd = current.second;

            mp[hd] = node->data;

            if (node->left) q.push({node->left, hd - 1});
            if (node->right) q.push({node->right, hd + 1});
        }

        for (auto const& [hd, val] : mp) {
            result.push_back(val);
        }

        return result;
    }
};
